solver/llbsolver/bridge.go

```diff
@@ -158,17 +161,60 @@ func (b *llbBridge) Solve(ctx context.Context, req frontend.SolveRequest, sid st
 		return &frontend.Result{}, nil
 	}
 
+	if res.Metadata == nil {
+		res.Metadata = make(map[string][]byte)
+	}
+
+	metadata := make(map[string][]byte)
+	for k, v := range res.Metadata {
+		if !strings.HasPrefix(k, exptypes.ExporterBuildInfo) {
+			metadata[k] = v
+			continue
+		}
+		bi := &binfotypes.BuildInfo{}
+		if errm := json.Unmarshal(v, bi); errm != nil {
+			return nil, errors.Wrapf(errm, "failed to unmarshal build info for %q", k)
+		}
+		if bi.Frontend == "" {
+			bi.Frontend = req.Frontend
+		}
+		bi.Attrs = buildinfo.ReduceMap(bi.Attrs, req.FrontendOpt)
+		dtbi, errm := json.Marshal(bi)
+		if errm != nil {
+			return nil, errm
+		}
+		metadata[k] = dtbi
+	}
+	if len(res.Refs) > 0 {
```

Comment:
if this was already found on line 169, then this shouldn't run I think. A better approach could be to not `range res.Metadata` at all but check if the exact key already exists and call `SetMetadata` or `ReduceMap` based on it (or combine to same function).

Reply:
Hum yeah we can combine them